btas/generic/cp_als.h

```diff
@@ -735,25 +746,27 @@ namespace btas{
                           Range1{pseudo_rank}});
         Tensor contract_tensor(Range{Range1{temp.extent(0)}, Range1{temp.extent(2)}});
         contract_tensor.fill(0.0);
+        const auto &a = A[(last_dim ? contract_dim + 1 : contract_dim)];
         // If the middle dimension is the mode not being contracted, I will move
         // it to the right hand side temp((size of tensor_ref/product of
         // dimension contracted, rank * mode n dimension)
         if (n == contract_dim) {
           pseudo_rank *= offset_dim;
         }
 
-          // If the code hasn't hit the mode of interest yet, it will contract
-          // over the middle dimension and sum over the rank.
+        // If the code hasn't hit the mode of interest yet, it will contract
+        // over the middle dimension and sum over the rank.
+
         else if (contract_dim > n) {
-          auto idx1 = temp.extent(0);
-          auto idx2 = temp.extent(1);
-          for(int i = 0; i < idx1; i++){
-            auto * contract_ptr = contract_tensor.data() + i * rank;
-            for(int j = 0; j < idx2; j++){
-              const auto * temp_ptr = temp.data() + i * idx2 * rank + j * rank;
-
-              const auto * A_ptr = A[(last_dim ? contract_dim + 1: contract_dim)].data() + j * rank;
-              for(int r = 0; r < rank; r++){
+          ord_t idx1 = temp.extent(0);
+          ord_t idx2 = temp.extent(1);
+          for (ord_t i = 0; i < idx1; i++) {
+            auto *contract_ptr = contract_tensor.data() + i * rank;
```

Comment:
it's better to rewrite these and other similar loops as 
```
ord_t i_times_rank = 0;
for (int_t i = 0; i < idx1; ++i, i_times_rank+=rank) {
  auto *contract_ptr = contract_tensor.data() + i_times_rank;
}
```
first, you trade multiply for addition (in many cases the latter will have lower latency and/or less power https://www.agner.org/optimize/). Second it scales to deeper loops like on line 784 where you are have multiple multiplies for no reason. Single add suffices.